### PR TITLE
Deadlock bug fixes (CU-8678wudz8)

### PR DESCRIPTION
**Pull request by tmackinnon**
Made some small changes to see if they will have any effect on the increased deadlock errors that have been coming in. I will monitor the Sentry error logs after these changes are deployed, then revisit to amend/update. Changes are outlined below and in the clickup task comments. I don't _think_ I need to add new tests for these changes, but let me know if you think otherwise.

**vitals.js**
- Added error handling to `sendLowBatteryAlert()` to stop subsequent db queries  

**BraveAlerterConfigurator.js**
- Added retry to `alertSessionChangedCallback()` per convo with Theresa 
- Retrying `beginTransaction` as we only have on shot at this function working (vs heartbeat that comes in often and can capture data the next time its called) 
- Retry has no time delay as of right now, but could be something to look into later

**db.js**
- Removed lock on tables that are not used in the functions that call beginTransaction
- Another solution could be to have `beginTransaction` take in a parameter that represents the exact tables needed to be locked - currently I don't think is necessary unless `beginTransaction` starts to gets used in more functions
- Added explicit return values to `beginTransaction` and `getMostRecentSensorsVitalWithLocationid`in their catch blocks
- Update parameter in `getMostRecentSensorsVitalWithLocationid()` so that it takes the whole `location` object INSTEAD of just `locationid` (since the location info is already available when this function gets called) . This would eliminate the need to call getLocations reducing the use of this db resource, and ideally reducing chance of deadlock.
   - Update the function name to `getMostRecentSensorsVitalWithLocation` to reflect this change

**checkHeartbeatTest.js**
- Update function name from `getMostRecentSensorsVitalWithLocationid` to `getMostRecentSensorsVitalWithLocation`

Test Plan:
- [x] Test case in `checkHeartbeatTest.js` passes: Error is logged from the `SendLowBatteryAlert` function when `db.beginTransaction` returns null 
- [x] Ensure Travis is passing
- [x] Deploy to Dev with no issues
- [x] Smoke test passes

### CODE REVIEW

File: server/vitals.js [hunk 1]
```diff
@@ -162,6 +162,10 @@ async function sendLowBatteryAlert(locationid) {
     const timeoutInMillis = parseInt(helpers.getEnvVar('LOW_BATTERY_ALERT_TIMEOUT'), 10) * 1000
 
     pgClient = await db.beginTransaction()
+    if (pgClient === null) {
+      helpers.logError(`sendLowBatteryAlert: Error starting transaction`)
```
Comment (trisapeace) [inline server/vitals.js:166]: You could add a test to make sure that this error log happens when `pgClient` is `null`.

**Review by trisapeace — APPROVED** [review]
Fantastic.

I have hope that this will make a real difference in the deadlocks!

I made some comments and suggestions, but nothing serious. You're choice if you make any of the changes.

File: server/BraveAlerterConfigurator.js [hunk 2]
```diff
@@ -90,8 +90,12 @@ class BraveAlerterConfigurator {
     try {
       pgClient = await db.beginTransaction()
       if (pgClient === null) {
-        helpers.logError(`alertSessionChangedCallback: Error starting transaction`)
-        return
+        // If the beginTransaction function failed to start the transaction, retry it once more before logging an error
+        pgClient = await db.beginTransaction()
+        if (pgClient === null) {
+          helpers.logError(`alertSessionChangedCallback: Error starting transaction`)
+          return
```
[the remark this answers is not in the document]
Comment (tmackinnon) [reply server/BraveAlerterConfigurator.js:97]: Oh that's a good point! Are you thinking in the catch block of `beginTransaction`?

File: server/db/db.js [hunk 3]
```diff
@@ -65,12 +65,13 @@ async function beginTransaction() {
 
     await pgClient.query('BEGIN')
 
-    await pgClient.query('LOCK TABLE clients, notifications, sessions, locations, sensors_vitals, sensors_vitals_cache')
```
[the remark this answers is not in the document]
Comment (tmackinnon) [reply server/db/db.js:68]: Ah yes that makes a lot of sense. And yes we definitely can revisit the idea of passing table names in as parameters in the future, if we find it necessary.

File: server/BraveAlerterConfigurator.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (trisapeace) [reply server/BraveAlerterConfigurator.js:97]: Yes, that was what I was thinking.

File: server/db/db.js [hunk 4]
```diff
@@ -1176,10 +1177,11 @@ async function getMostRecentSensorsVitalWithLocationid(locationid, pgClient) {
       return null
     }
 
-    const allLocations = await getLocations(pgClient)
+    const allLocations = [location]
     return createSensorsVitalFromRow(results.rows[0], allLocations)
```
[the remark this answers is not in the document]
Comment (tmackinnon) [reply server/db/db.js:1181]: You're so right, will be much clearer that way!

File: server/db/db.js [hunk 5]
```diff
@@ -67,6 +68,11 @@ async function beginTransaction() {
 
     await pgClient.query('LOCK TABLE clients, sessions, locations')
   } catch (e) {
+    if (!isRetry) {
+      isRetry = true
```
Comment (trisapeace) [inline server/db/db.js:72]: Hmmm, does this work? I would have expected `isRetry` to be `false` again each time you call `beginTransaction`.

Comment (tmackinnon) [reply server/db/db.js:72]: Hmm good point, I think instead I'll add a default parameter for the retryCount then use number for my conditional. Stay tuned!

File: server/vitals.js (hunk 1 above)
Comment (tmackinnon) [reply server/vitals.js:166]: Good idea, added!

File: server/db/db.js [hunk 6]
```diff
@@ -1176,10 +1181,10 @@ async function getMostRecentSensorsVitalWithLocationid(locationid, pgClient) {
       return null
     }
 
-    const allLocations = await getLocations(pgClient)
-    return createSensorsVitalFromRow(results.rows[0], allLocations)
+    return createSensorsVitalFromRow(results.rows[0], [location])
```
Comment (micahdbak) [inline server/db/db.js:1184]: Very nice change here. Removes an unnecessary (and probably slow) query to get all locations; great catch.

File: server/test/integration/vitalsTest/handleHeartbeatTest.js [hunk 7]
```diff
@@ -9,6 +9,7 @@ const sinon = require('sinon')
 const { factories, helpers } = require('brave-alert-lib')
 const { braveAlerter, db, server } = require('../../../index')
 const { locationDBFactory, sensorsVitalDBFactory } = require('../../../testingHelpers')
+// const { sendLowBatteryAlert } = require('../../../vitals')
```
Comment (micahdbak) [inline server/test/integration/vitalsTest/handleHeartbeatTest.js:12]: Is there a reason that this line is only commented out, as opposed to removed?

**Review by micahdbak — APPROVED** [review]
Looks good! Looks like what you've changed is pretty small, but I hope that makes an impact on the deadlocks. I look forward to seeing what results from this.

File: server/test/integration/vitalsTest/handleHeartbeatTest.js (hunk 7 above)
Comment (tmackinnon) [reply server/test/integration/vitalsTest/handleHeartbeatTest.js:12]: Oh whoops thank you for catching that! That line shouldn't be in there. I thought I would maybe need it when I was writing the test but I didn't and I guess forgot to delete it. Thanks!!

File: server/db/db.js (hunk 6 above)
Comment (tmackinnon) [reply server/db/db.js:1184]: Thanks!
